cmd/maya-exporter/app/collector/collector.go

```diff
@@ -0,0 +1,155 @@
+package collector
+
+import (
+	"github.com/golang/glog"
+	"github.com/openebs/maya/types/v1"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// collector inherits the properties of cstor and jiva,
+// these properties includes metrics of the volumes.
+type collector struct {
+	Volume
+	metrics
+}
+
+func New(vol Volume, casType string) *collector {
+	return &collector{
+		vol,
+		Metrics(casType),
+	}
+}
+
+// gaugeList returns the list of the registered gauge variables
+func (c *collector) gaugesList() []prometheus.Gauge {
+	return []prometheus.Gauge{
+		c.reads,
+		c.writes,
+		c.totalReadBytes,
+		c.totalWriteBytes,
+		c.totalReadTime,
+		c.totalWriteTime,
+		c.totalReadBlockCount,
+		c.totalWriteBlockCount,
+		c.actualUsed,
+		c.logicalSize,
+		c.sectorSize,
+		c.sizeOfVolume,
+		c.volumeStatus,
+	}
+}
+
+// counterList returns the list of registered counter variables
+func (c *collector) countersList() []prometheus.Collector {
+	return []prometheus.Collector{
+		c.volumeUpTime,
+		c.connectionErrorCounter,
+		c.connectionRetryCounter,
+		c.totalReplicaCounter,
+		c.degradedReplicaCounter,
+		c.healthyReplicaCounter,
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once
+// the last descriptor has been sent. The sent descriptors fulfill the
+// consistency and uniqueness requirements described in the Desc
+// documentation. (It is valid if one and the same Collector sends
+// duplicate descriptors. Those duplicates are simply ignored. However,
+// two different Collectors must not send duplicate descriptors.) This
+// method idempotently sends the same descriptors throughout the
+// lifetime of the Collector. If a Collector encounters an error while
+// executing this method, it must send an invalid descriptor (created
+// with NewInvalidDesc) to signal the error to the registry.
+
+// Describe implements Describe method of prometheus.Collector interface.
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	for _, gauge := range c.gaugesList() {
+		gauge.Describe(ch)
+	}
+
+	for _, counter := range c.countersList() {
+		counter.Describe(ch)
+	}
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent. The
+// descriptor of each sent metric is one of those returned by
+// Describe. Returned metrics that share the same descriptor must differ
+// in their variable label values. This method may be called
```

Comment:
Is it possible to provide example (in the comments) on when this stuff becomes duplicate or so.